flag.go

```diff
@@ -85,35 +92,56 @@ func bucketUser(user User, key, attr, salt string) float32 {
 }
 
 func (f FeatureFlag) EvaluateExplain(user User) (interface{}, *Explanation) {
+	index, explanation := f.evaluateExplainIndex(user)
+
+	if index == nil || *index >= len(f.Variations) {
+		return nil, explanation
+	} else {
+		return f.Variations[*index].Value, explanation
+	}
+
+}
+
+func (f FeatureFlag) evaluateExplainIndex(user User) (*int, *Explanation) {
 	// TODO: The toggle algorithm should check the kill switch. We won't check it here
 	// so we can potentially compute an explanation even if the kill switch is hit
 	if user.Key == nil {
 		return f.OffVariation, nil
 	}
 
+	// Check to see if targets match
 	for _, target := range f.Targets {
 		for _, value := range target.Values {
 			if value == *user.Key {
 				explanation := Explanation{Kind: "target", Target: &target}
-				return target.Variation, &explanation
+				return &target.Variation, &explanation
 			}
 		}
 	}
 
+	// Now walk through the rules and see if any match
 	for _, rule := range f.Rules {
 		if rule.matchesUser(user) {
-			variation, passErr := rule.valueForUser(user, f.Key, f.Salt)
+			variation := rule.variationIndexForUser(user, f.Key, f.Salt)
 
-			if passErr {
-				return f.OffVariation, nil
+			if variation == nil {
+				return f.OffVariation, nil // TODO should this continue, or return the off variation?
```

Comment:
This is a good question. What is the expected behavior when multiple rules match a user? first one wins? if so then we return here.
 
I think this is the case I was asking about in the FF2 meeting. Someone makes a rule including gmail.com users then another rule excluding gmail.com users. Is this example correct?